src/envoy/auth/authenticator.h

```diff
@@ -0,0 +1,85 @@
+/* Copyright 2017 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef AUTHENTICATOR_H
+#define AUTHENTICATOR_H
+
+#include "common/common/logger.h"
+#include "envoy/http/async_client.h"
+#include "server/config/network/http_connection_manager.h"
+
+#include "src/envoy/auth/auth_store.h"
+
+namespace Envoy {
+namespace Http {
+namespace Auth {
+
+// A per-request JWT authenticator to handle all JWT authentication:
+// * fetch remote public keys and cache them.
+class Authenticator : public Logger::Loggable<Logger::Id::http>,
```

Comment:
I would call this JwtAuthenticator to avoid possible collision with other authenticator in the future.